friends/services/friends_services.py

```diff
@@ -86,3 +94,80 @@ def get_all_received_friend_requests(user_id: str) -> Optional[ExportFriendReque
             return all_friend_details
         else:
             return None
+
+    @staticmethod
+    def save_friend_request(sender: User, receiver: User) -> FriendRequest:
+        new_friend_request = FriendRequest(sender=sender, receiver=receiver)
+        new_friend_request.save()
+        return new_friend_request
+
+    @staticmethod
+    def create_new_friend_request_service(
+        request_data: AddFriendRequestType, uid: str
+    ) -> dict:
+        is_validate_request: bool = FriendSerializer().validate(
+            data=request_data.model_dump()
+        )
+        if is_validate_request:
+            # Retrieve sender and receiver
+            receiver: User = User.objects.get(email=request_data.user_email)
+            sender: User = User.objects.get(id=uid)
+
+            already_sent_request: bool = FriendRequest.objects.filter(
+                sender=sender, receiver=receiver
+            ).exists()
+
+            reversed_sent_request: bool = FriendRequest.objects.filter(
+                sender=receiver, receiver=sender
+            ).exists()
+
+            already_a_friend: bool = Friend.objects.filter(
+                Q(user1=sender, user2=receiver) | Q(user1=receiver, user2=sender)
+            ).exists()
+
+            is_self_user: bool = True if sender.id == receiver.id else False
+
+            receiver_info: str = (
+                receiver.username.upper() if receiver.username else receiver.email
+            )
+
+            # sender cant be receiver
+            if is_self_user:
+                return {
+                    "successMessage": None,
+                    "errorMessage": "You can't send request to yourself",
+                }
+
+            # Check if a friend request already exists
+            if already_sent_request:
+                return {
+                    "successMessage": None,
+                    "errorMessage": f"Friend request already sent to this {receiver_info}.",
+                }
+
+            # Check if the reverse friend request exists (receiver has sent to sender)
+            if reversed_sent_request:
+                return {
+                    "successMessage": None,
+                    "errorMessage": f"{receiver_info} has already sent you a friend request.",
+                }
+
+            # Check if they are already friends
+            if already_a_friend:
+                return {
+                    "successMessage": None,
+                    "errorMessage": f"You are already friends with this {receiver_info}.",
+                }
+
```

Comment:
These all should be handled with exception handling.

Reply:
exceptions are handled